File: bukkit/freezer/src/main/java/dialight/freezer/FreezerListener.java
package dialight.freezer;

import dialight.extensions.PlayerEx;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockDamageEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.block.SignChangeEvent;
import org.bukkit.event.enchantment.EnchantItemEvent;
import org.bukkit.event.enchantment.PrepareItemEnchantEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.FoodLevelChangeEvent;
import org.bukkit.event.hanging.HangingBreakByEntityEvent;
import org.bukkit.event.hanging.HangingPlaceEvent;
import org.bukkit.event.inventory.CraftItemEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.event.player.*;
import org.bukkit.event.server.PluginDisableEvent;
import org.bukkit.event.vehicle.VehicleEnterEvent;
import org.bukkit.event.vehicle.VehicleExitEvent;
import org.bukkit.plugin.PluginManager;

public class FreezerListener implements Listener {

    private final Freezer proj;
    private boolean enabled = false;

    public FreezerListener(Freezer proj) {
        this.proj = proj;
    }

    public void enable() {
        if(enabled) return;
        PluginManager pluginManager = proj.getPlugin().getServer().getPluginManager();
        pluginManager.registerEvents(this, proj.getPlugin());
        enabled = true;
    }

    public void disable() {
        if(!enabled) return;
        HandlerList.unregisterAll(this);
        enabled = false;
    }

    @EventHandler public void onTeleport(PlayerTeleportEvent e) {
        Frozen frozen = proj.get(e.getPlayer().getUniqueId());
        if(frozen != null) {
            frozen.setLocation(e.getTo());
        }
    }

    @EventHandler public void onMove(PlayerMoveEvent e) {
        Player trg = e.getPlayer();
        Frozen frozen = this.proj.get(trg.getUniqueId());
        if(frozen == null) return;
        if(!blockMoved(e.getFrom(), e.getTo())) return;
        PlayerEx.of(trg).teleport(frozen.getLocation());
        if(frozen.getLocation().clone().add(0, -1, 0).getBlock().getType().isBlock()) {
            proj.getFreezeFlayers().removeFly(frozen.getTarget());
        } else {
            proj.getFreezeFlayers().setFly(frozen.getTarget());
        }
    }

    private boolean blockMoved(Location f, Location t) {
        if(f.getBlockX() != t.getBlockX()) return true;
        if(f.getBlockY() != t.getBlockY()) return true;
        if(f.getBlockZ() != t.getBlockZ()) return true;
        return false;
    }

    @EventHandler public void onPluginDisable(PluginDisableEvent e) {
        if (!e.getPlugin().getName().equals(proj.getPlugin().getName())) return;
        for (Frozen frozen : proj.getFrozens()) {
            Player trg = Bukkit.getPlayer(frozen.getTarget().getUuid());
            if(trg == null) continue;
            trg.sendMessage(FreezerMessages.unfreezeByReload);
        }
    }

    @EventHandler public void onJoin(PlayerJoinEvent e) {
        Player trg = e.getPlayer();
        Frozen frozen = this.proj.get(trg.getUniqueId());
        if(frozen == null) return;
        PlayerEx.of(trg).teleport(frozen.getLocation());
        trg.sendMessage(FreezerMessages.youFrozen);
    }
    @EventHandler public void onQuit(PlayerQuitEvent e) {

    }

    //////////////////////////////HOLDER/////////////////////////////////

    //ломать и ставить блоки
    @EventHandler public void onBlockBreak(BlockBreakEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }
    @EventHandler public void onBlockPlace(BlockPlaceEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //садиться на шифт
    @EventHandler public void toggleSneak(PlayerToggleSneakEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.getPlayer().setSneaking(false);
        e.setCancelled(true);
    }

    //наносить и получать урон
    @EventHandler public void onDamage(EntityDamageEvent e) {
        if(e.getEntity().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getEntity().getUniqueId());
        if(frozen == null) return;
        e.setDamage(0d);
        e.setCancelled(true);
    }
    @EventHandler public void onDealDamage(EntityDamageByEntityEvent e) {
        if(e.getDamager().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getDamager().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //выкидывать и подбирать вещи
    @EventHandler public void onDrop(PlayerDropItemEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }
    @EventHandler public void onPickup(PlayerPickupItemEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //использовать транспорт
    @EventHandler public void enterVehicle(VehicleEnterEvent e) {
        if(e.getEntered().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getEntered().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }
    @EventHandler public void exitVehicle(VehicleExitEvent e) {
        if(e.getExited().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getExited().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //выливать и набирать жидкости в ведра
    @EventHandler public void onEmpty(PlayerBucketEmptyEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }
    @EventHandler public void onFill(PlayerBucketFillEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //использовать кровать
    @EventHandler
    public void onEnterBed(PlayerBedEnterEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    //открывать и использовать инвентарь
    @EventHandler public void onInventory(InventoryClickEvent e) {
        if(e.getWhoClicked().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getWhoClicked().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }
    @EventHandler public void onOpen(InventoryOpenEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        if(e.getPlayer().isOp()) return;
        e.setCancelled(true);
    }

    @EventHandler public void onFoodLevelChange(FoodLevelChangeEvent e) {
        if(e.getEntityType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getEntity().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void sign(SignChangeEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void blockDamage(BlockDamageEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void enchantItem(EnchantItemEvent e) {
        Frozen frozen = this.proj.get(e.getEnchanter().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onPrepareEnchant(PrepareItemEnchantEvent e) {
        Frozen frozen = this.proj.get(e.getEnchanter().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onBreakHanging(HangingBreakByEntityEvent e) {
        if(e.getRemover().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getRemover().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onPlaceHanging(HangingPlaceEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onCraft(CraftItemEvent e) {
        if(e.getWhoClicked().getType() != EntityType.PLAYER) return;
        Frozen frozen = this.proj.get(e.getWhoClicked().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onFish(PlayerFishEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onItemConsume(PlayerItemConsumeEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onShear(PlayerShearEntityEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

    @EventHandler public void onInteractEntity(PlayerInteractEntityEvent e) {
        Frozen frozen = this.proj.get(e.getPlayer().getUniqueId());
        if(frozen == null) return;
        e.setCancelled(true);
    }

}

File: src/main/freezer/dialight/freezer/FreezerListener.kt
package dialight.freezer

import dialight.extensions.dump
import dialight.teleporter.PlayerTeleportEvent
import org.spongepowered.api.Sponge
import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.event.Listener
import org.spongepowered.api.event.cause.Cause
import org.spongepowered.api.event.cause.EventContext
import org.spongepowered.api.event.entity.MoveEntityEvent
import org.spongepowered.api.event.filter.cause.First
import org.spongepowered.api.text.Text
import org.spongepowered.api.world.Location
import org.spongepowered.api.world.World

class FreezerListener(val plugin: FreezerPlugin) {

    @Listener
    fun onTeleport(event: PlayerTeleportEvent, @First player: Player) {
        event.dump()
        val frozen = plugin.freezer.frozen.idFrozen[player.uniqueId] ?: return
        frozen.setLocation(event.getTo())
    }

    @Listener
    fun onMove(e: MoveEntityEvent, @First trg: Player) {
        val frozen = plugin.freezer.frozen.idFrozen[trg.uniqueId] ?: return
        if (!blockMoved(e.fromTransform.location, e.toTransform.location)) return
//        this.teleportFrozen(frozen, trg)
    }

    private fun blockMoved(f: Location<World>, t: Location<World>): Boolean {
        if (f.blockX != t.blockX) return true
        if (f.blockY != t.blockY) return true
        return f.blockZ != t.blockZ
    }

//    @EventHandler
//    fun onPluginDisable(e: PluginDisableEvent) {
//        if (e.getPlugin().getName() != eventHelper.plugin.getName()) return
//        for (frozen in this.idFrozen.values) {
//            val trg = Bukkit.getPlayer(frozen.getUniqueId()) ?: continue
//            FreezerMessages.unfreezeByReload(trg)
//        }
//    }
//
//    @EventHandler
//    fun onJoin(e: PlayerJoinEvent) {
//        val trg = e.getPlayer()
//        val frozen = this.offlineFrozen.remove(trg.getUniqueId()) ?: return
//        frozen.updateAll(trg)
//        this.idFrozen.put(trg.getEntityId(), frozen)
//        this.onFreezeOnline(frozen, trg)
//        FreezerMessages.frozen(trg)
//    }
//
//    @EventHandler
//    fun onQuit(e: PlayerQuitEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        this.offlineFrozen.put(e.getPlayer().getUniqueId(), frozen)
//    }
//
//    protected fun teleportFrozen(frozen: Frozen, trg: Player) {
//        trg.teleport(Utils.tpFix(frozen.getLocation(), trg.getLocation()))
//    }
//
//    protected override fun onFreezeOnline(frozen: Frozen, trg: Player) {
//        super.onFreezeOnline(frozen, trg)
//        this.teleportFrozen(frozen, trg)
//        this.freezeFlayers.setFly(trg)
//    }
//
//    protected override fun onUnfreezeOnline(frozen: Frozen, trg: Player) {
//        super.onUnfreezeOnline(frozen, trg)
//        this.freezeFlayers.removeFly(trg)
//        frozen.recoverVelocity(trg)
//    }
//
//    protected override fun onFreezeOffline(frozen: Frozen) {
//        super.onFreezeOffline(frozen)
//        this.freezeFlayers.setFlyFuture(frozen.getUniqueId())
//    }
//
//    protected override fun onUnfreezeOffline(frozen: Frozen) {
//        super.onUnfreezeOffline(frozen)
//        this.freezeFlayers.removeFlyFuture(frozen.getUniqueId())
//    }
//
//    //////////////////////////////HOLDER/////////////////////////////////
//
//    //ломать и ставить блоки
//    @EventHandler
//    fun onBlockBreak(e: BlockBreakEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onBlockPlace(e: BlockPlaceEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //садиться на шифт
//    @EventHandler
//    fun toggleSneak(e: PlayerToggleSneakEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.getPlayer().setSneaking(false)
//        e.setCancelled(true)
//    }
//
//    //наносить и получать урон
//    @EventHandler
//    fun onDamage(e: EntityDamageEvent) {
//        if (e.getEntity().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getEntity().getEntityId()) ?: return
//        e.setDamage(0.0)
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onDealDamage(e: EntityDamageByEntityEvent) {
//        if (e.getDamager().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getDamager().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //выкидывать и подбирать вещи
//    @EventHandler
//    fun onDrop(e: PlayerDropItemEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onPickup(e: PlayerPickupItemEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //использовать транспорт
//    @EventHandler
//    fun enterVehicle(e: VehicleEnterEvent) {
//        if (e.getEntered().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getEntered().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun exitVehicle(e: VehicleExitEvent) {
//        if (e.getExited().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getExited().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //выливать и набирать жидкости в ведра
//    @EventHandler
//    fun onEmpty(e: PlayerBucketEmptyEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onFill(e: PlayerBucketFillEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //использовать кровать
//    @EventHandler
//    fun onEnterBed(e: PlayerBedEnterEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    //открывать и использовать инвентарь
//    @EventHandler
//    fun onInventory(e: InventoryClickEvent) {
//        if (e.getWhoClicked().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getWhoClicked().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onOpen(e: InventoryOpenEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onFoodLevelChange(e: FoodLevelChangeEvent) {
//        if (e.getEntityType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getEntity().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun sign(e: SignChangeEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun blockDamage(e: BlockDamageEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun enchantItem(e: EnchantItemEvent) {
//        val frozen = this.idFrozen.get(e.getEnchanter().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onPrepareEnchant(e: PrepareItemEnchantEvent) {
//        val frozen = this.idFrozen.get(e.getEnchanter().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onBreakHanging(e: HangingBreakByEntityEvent) {
//        if (e.getRemover().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getRemover().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onPlaceHanging(e: HangingPlaceEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onCraft(e: CraftItemEvent) {
//        if (e.getWhoClicked().getType() != EntityType.PLAYER) return
//        val frozen = this.idFrozen.get(e.getWhoClicked().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onFish(e: PlayerFishEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onItemConsume(e: PlayerItemConsumeEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onShear(e: PlayerShearEntityEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }
//
//    @EventHandler
//    fun onInteractEntity(e: PlayerInteractEntityEvent) {
//        val frozen = this.idFrozen.get(e.getPlayer().getEntityId()) ?: return
//        e.setCancelled(true)
//    }



}